README.md
# MMM-AC-aseag

This is a module for the [MagicMirror²](https://github.com/MichMich/MagicMirror/).
It display the realtime departures of busses in Aachen using the API provided by the ASEAG. The module is still work in progress.  

This is my first module for the MagicMirror and my first project in js, so feel free to contribute or comment.


## Installation
1. Navigate into your MagicMirror's `modules` folder and excecute `git clone https://github.com/neandertaler19/MMM-AC-aseag`
2. Navigate into the newly created folder.
3. Execute `npm install`to install the node dependencies.

## Using the module

To use this module, add the following configuration block to the modules array in the `config/config.js` file:
```js
{
    module: "MMM-AC-aseag",
    position: "bottom_left",
    header: "Misereor Abfahrten",
    config: {
        retryDelay: 5 * 1000, // Time to wait before retrying in case of failure
        maximumEntries: 10, // Maximum number of entries shown in table
        updateInterval: 1 * 60 * 1000, // Update interval in microseconds
        stop: "",
    }
},

```

## Configuration options

| Option           | Description
|----------------- |-----------
| `retryDelay`        | *optional* TIme to wait before retrying in case of failure
| `maximumEntries`| *optional* Maximum number of entries shown in the table
| `updateInterval`        | *optional* Update interval for querying for new connections <br><br>**Type:** `int`(milliseconds) <br>Default 60000 milliseconds (1 minute)
| `stop` | *optional* name of the stop. If you don't provide this information, you should leave it empty (`""`) and provide a stopID instead
| `stopID` | *optional* ID of the stop


## Acknowledgements
* [busplan](https://github.com/johschmitz/busplan)
* [MMM-HH-localtransport](https://github.com/georg90/MMM-HH-LocalTransport)
* [MMM-DWD-WarnWeather](https://github.com/LukeSkywalker92/MMM-DWD-WarnWeather)


CHANGELOG.md
# MMM-AC-aseag Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).


## [1.0.0] - Initial release

First public release


node_helper.js
/* Magic Mirror
 * Node Helper: MMM-AC-aseag
 *
 * By neandertaler19
 * MIT Licensed.
 */

var NodeHelper = require("node_helper");
var request = require("request");

module.exports = NodeHelper.create({

	updateTimetable: function(stopID) {
		var self = this;
		// URL for querying the bus stop departures
		var urlApi = "http://ivu.aseag.de/interfaces/ura/instant_V2?ReturnList=LineName,DestinationText,EstimatedTime&StopID=" + stopID;

		// Send a request for the departures
		request({
			url: urlApi,
			method: "GET"
		}, function(error, response, body) {
			// Repair the JSON
			var str = "[" + body.split("\n").slice(1).join(",") + "]";
			// Parse the result
			var jsonObj = JSON.parse(str);
			// Sort the array by the departure
			jsonObj.sort(function(a, b) {
				if (a[3] == b[3]) {
					return 0;
				}
				if (a[3] < b[3]) {
					return -1;
				}
				if (a[3] > b[3]) {
					return 1;
				}
			});
			// Send it back to the main function
			self.sendSocketNotification("BUSSES", jsonObj);
		});
	},

	extractID: function(name) {
		var self = this;
		// URL for querying the ID for the stop name
		var urlLoc = "http://ivu.aseag.de/interfaces/ura/location?maxResults=10&searchTypes=STOPPOINT&searchString=" + name;

		// Send a request for the stop name
		request({
			url: urlLoc,
			method: "GET"
		}, function(error, repsonse, body) {
			// Prase the result
			jsonObj = JSON.parse(body);
			// Read out the stop name
			var ID = jsonObj["resultList"][0]["stopPointId"];
			// Send the name back to main function
			self.sendSocketNotification("RETURN_ID", ID);
		});
	},

	// Override socketNotificationReceived method.

	/* socketNotificationReceived(notification, payload)
	 * This method is called when a socket notification arrives.
	 *
	 * argument notification string - The identifier of the noitication.
	 * argument payload mixed - The payload of the notification.
	 */
	socketNotificationReceived: function(notification, payload) {
		const self = this;
		if (notification == "GET_BUSSES") {
			self.updateTimetable(payload);
		}
		if (notification == "GET_ID") {
			self.extractID(payload);
		}
	},
});


MMM-AC-aseag.js
/* global Module */

/* Magic Mirror
 * Module: MMM-AC-aseag
 *
 * By neandertaler19
 * MIT Licensed.
 */

Module.register("MMM-AC-aseag", {
	defaults: {
		retryDelay: 5000,	// Delay for retrying
		maximumEntries: 10, // Maximum entries shown
		updateInterval: 1 * 60 * 1000, // Update every minute.
		stop: "Misereor",	// The name of the stop. If you leave this empty, the ID will be used
		stopID: 100010, // Stop for displaying the departures
	},

	start: function() {
		Log.info("Starting module: " + this.name);
		// Initialize the bus array
		this.busses = [];
		this.loaded = false;
		this.updateTimer = null;
		// If a name is provided, get it's ID
		if (this.config.stop != "") {
			this.sendSocketNotification("GET_ID", this.config.stop);
		} else {
			// If not, directly use the ID and update the departures
			this.updateBusses(this);
		}
	},

	updateBusses: function (self) {
		// Call the helper to get the busses
		self.sendSocketNotification('GET_BUSSES', self.config.stopID);
		// Configure it to be called every *updateInterval* minute
		setTimeout(self.updateBusses, self.config.updateInterval, self);		
	},

	getDom: function() {
		// create element wrapper for show into the module
		var wrapper = document.createElement("div");

		// The ID is too low to be correct
		if (this.config.id < 100000) {
			wrapper.innerHTML = "The ID " +this.id + " provided is not correct";
			wrapper.className = "dimmed light small";
			return wrapper;
		}

		// Display text while loading the connections
		if (!this.loaded) {
			wrapper.innerHTML = "Loading connections ...";
			wrapper.className = "dimmed light small";
			return wrapper;
		}

		// Display the departures in a table
		var table = document.createElement("table");
		table.className = "small";
		// The current time
		var current = Date.now();

		for (var b in this.busses) {
			if (b > this.config.maximumEntries) {
				break;
			}
			// Define the subarray for quick access
			var bus = this.busses[b];
			var departure = new Date(bus[3]).getTime();
			var remaining = Math.round((departure - current)/1000/60);

			// Create a new row in the table
			var row = document.createElement("tr");
			table.appendChild(row);

			// Fill it with the line number
			var lineCell = document.createElement("td");
			lineCell.className = "align-right bright";
			lineCell.innerHTML = bus[1];
			row.appendChild(lineCell);

			// Fill it with the destination
			var toCell = document.createElement("td");
			toCell.classname = "align-right trainto"
			toCell.innerHTML = bus[2];
			row.appendChild(toCell);

			// Fill it with the departure time
			var depCell = document.createElement("td");
			depCell.className = "departuretime";
			if (remaining <= 0){
				depCell.innerHTML = "now";
			} else {
				depCell.innerHTML = remaining + " min";
			}
			row.appendChild(depCell);
		}
		return table;
	},

	getStyles: function () {
		return ["MMM-AC-aseag.css"];
	},

	// socketNotificationReceived from helper
	// Not used at the moment
	socketNotificationReceived: function (notification, payload) {
		if (notification == "BUSSES") {
			this.busses = payload;
			this.loaded = true;
			this.updateDom(this.config.animationSpeed);
		}
		if (notification == "RETURN_ID") {
			this.config.stopID = payload;
			Log.info("Stop ID: " + payload);
			this.updateBusses(this);
		}
	},
});
